cpp/src/fil/infer.cu

```diff
@@ -236,19 +238,35 @@ struct tree_aggregator_t {
     : tmp_storage(finalize_workspace) {}
 
   __device__ __forceinline__ void accumulate(
-    vec<NITEMS, float> single_tree_prediction, int tree, int num_rows) {
+    vec<NITEMS, float> single_tree_prediction, int tree) {
     acc += single_tree_prediction;
   }
 
-  __device__ __forceinline__ void finalize(float* out, int num_rows,
+  __device__ __forceinline__ void finalize(float* block_out, int block_num_rows,
                                            int output_stride,
-                                           output_t transform, int num_trees) {
-    __syncthreads();
-    acc = block_reduce(acc, vectorized(cub::Sum()), tmp_storage);
-    if (threadIdx.x > 0) return;
+                                           output_t transform, int num_trees,
+                                           int log2_threads_per_tree) {
+    if (FIL_TPB != 1 << log2_threads_per_tree) {  // anything to reduce?
+      // ensure input columns can be overwritten (no threads traversing trees)
+      __syncthreads();
+      if (log2_threads_per_tree == 0) {
+        acc = block_reduce(acc, vectorized(cub::Sum()), tmp_storage);
+      } else {
+        auto per_thread = (vec<NITEMS, float>*)tmp_storage;
+        per_thread[threadIdx.x] = acc;
+        __syncthreads();
+        acc = multi_sum<5>(per_thread, 1 << log2_threads_per_tree,
```

Comment:
Please add a comment explaining the choice of this particular template parameter value (e.g. because it gives the best performance, or just good performance).

For the future, consider having a version of multi-sum with this particular parameter hard-coded to some good default value.

Reply:
the best value depends on the expected size of `num_groups`. The bigger the `num_groups`, the bigger value becomes optimal. Binary or quaternary reduction is a good default value, but we have decided it's not sufficiently generic, and indeed, there are gains to be made from this fine-tuning.

Reply:
I decided to benchmark, and in this case, somehow the radix doesn't matter at all (~1% difference at best). I can think of a way to benchmark this and make a rule of thumb.